### ISSUE TITLE
[Feature]: Add support for layers on maps widget

### ISSUE THREAD

**Issue by SatishGandham** [appsmithorg/appsmith#9309]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Add support for layers on maps widget.
Layers are a container for geo spatial data. You can display any kind of geo spatial data by passing it in json format.

https://developers.google.com/maps/documentation/javascript/layers

### Why should this be worked on?

It's a common use case.

**Comment by dilippitchika** [comment]
Hey @SatishGandham is this a community request?

**Comment by rfseancavanaugh** [comment]
I'm also hoping for this feature!

**Comment by somangshu** [comment]
We need a POC on this

**Comment by momcilo-appsmith** [comment]
## Stats
| Stat | Values |
| ------------- | ------------- |
| Reach     | 63 |
| Effort (months)  | 0.75 |

**Comment by patrickswelsh** [comment]
Hey - I wanted to bump this / signal my interest in this request. I'm particularly interested in Google's [Data-Driven Styling for Boundaries.](https://developers.google.com/maps/documentation/javascript/dds-boundaries/overview). So perhaps that's a separate request, but I thought I'd start here.

I want to be able to display these boundaries and give the user the option to select specific regions (vs. the current point-based selection option). [This is an example](https://developers.google.com/maps/documentation/javascript/dds-boundaries/handle-events) of this done in a basic react script.

I'm new to contributing to Appsmith, but I'm totally willing to throw some hours into developing this feature. I've [made a fork ](https://github.com/patrickswelsh/appsmith_mapselect) and been fiddling with it, but I'm having a lot of trouble. I've replicated the example I cited above as a standalone script in my own environment. However, I haven't been able to translate that to Appsmith. I haven't even been able to display the boundaries non-interactively.

Is there a write-up somewhere of how the Map widget code works? Or someone who's willing to explain to a newbie?

**Comment by somangshu** [comment]
Hey @patrickswelsh, thanks for your interest here.
Map chart is for representing statistical data on maps. ex: country population etc. You can find more in the [docs](https://docs.appsmith.com/reference/widgets/map-chart#docusaurus_skipToContent_fallback)
On this issue, we haven’t been able to prioritise it yet. Simply because we haven’t seen much of a request around it. 
However, I found that this example you are talking about can be embedded and hence I tried [this](https://release.app.appsmith.com/app/untitled-application-11/page5-646382d0d4c3b50a6b27b34e) out and it worked! In the given example I am using the iframe widget to embed the link given on the google dev doc you shared. 

Basically if you are able to house your widget in some code sandbox, you can further use `onMessageReceived` event to talk to it and take care of advance controls.

cc @GreenFlux

**Comment by patrickswelsh** [comment]
@somangshu - Wow, time really got away from me here. I want to say - THANK YOU for this comment.

I took your suggestion and ran with it. I'm fairly new to frontend dev so it took me sme time to understand everything. I wanted to leave a note here describing what I did in case anyone else finds this thread and is interested in doing something similar:

Basically - Appsmith's ability to interact with iframes means that if you want some functionality that isn't currently built into Appsmith, you can develop and host the bit of code you need on a separate server and embed it in your Appsmith app. Appsmith has a great documentation on this [here](https://docs.appsmith.com/reference/appsmith-framework/widget-actions/post-message), and there's a video walkthrough [here](https://www.youtube.com/live/bccTy3v6dSE?feature=share&t=455).

It all comes down to this bit of javascript: window.parent.postMessage(). You can put that in your separately-hosted-custom-code-thingy to tell it to send a message out to appsmith when something happens. In my  case, I told it to send Appsmith the Postal Code of the area clicked on a map. On the appsmith side, you can use onMessageRecieved() to handle those messages.

I ended up hosting my bit of map code on Vercel, since they have a decent free tier and integrate nicely with github. Here's [my repo for reference.](https://github.com/patrickswelsh/map). Most of the files have to do with Vite, the framework I built off of, but the one where you can see my code for sending messages [is this file.](https://github.com/patrickswelsh/map/blob/main/index.ts)

Anyway - hope that helps the next newbie who stumbles on this thread. Thanks again, @somangshu.

**Comment by somangshu** [comment]
Thanks @patrickswelsh for taking the time to implement this and explaining it here. This will help new folks facing a similar challenge ❤️ Kudos to you 🚀
